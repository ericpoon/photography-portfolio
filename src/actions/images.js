import firebase from '../firebase/firebase';

const database = firebase.database();
const storage = firebase.storage();

export const setImages = images => ({
  type: 'SET_IMAGES',
  images,
});

export const startDownloadImages = () => {
  return (dispatch) => {
    return database.ref('/images').once('value')
      .then((snapshot) => {
        const images = [];
        snapshot.forEach((childSnapshot) => {
          images.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setImages(images));
      });
  };
};

export const addImage = image => ({
  type: 'ADD_IMAGE',
  image,
});

export const startAddImage = (image) => {
  const {
    title, subtitle, description, file,
  } = image;
  return (dispatch) => {
    const id = database.ref('/images').push().key;
    return storage.ref(`/images/${id}`).put(file, { cacheControl: 'private, max-age=60' })
      .then((snapshot) => {
        const fileUrl = snapshot.downloadURL;
        return database.ref(`/images/${id}`).set({
          title, subtitle, description, fileUrl,
        })
          .then(() => {
            dispatch(addImage({
              id,
              title,
              subtitle,
              description,
              fileUrl,
            }));
          });
      });
  };
};

export const editImage = (id, updates) => ({
  type: 'EDIT_IMAGE',
  id,
  updates,
});

export const startEditImage = (id, updates) => {
  return (dispatch) => {
    return database.ref(`/images/${id}`).update(updates)
      .then(() => {
        dispatch(editImage(id, updates));
      });
  };
};

export const deleteImage = id => ({
  type: 'DELETE_IMAGE',
  id,
});

export const startDeleteImage = (id) => {
  return (dispatch) => {
    return database.ref(`images/${id}`).remove()
      .then(() => {
        dispatch(deleteImage(id));
      });
  };
};
