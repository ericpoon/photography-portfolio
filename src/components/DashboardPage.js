import React from 'react';

export default () => (
  <div>
    <p>Dashboard page goes here</p>
  </div>
);
