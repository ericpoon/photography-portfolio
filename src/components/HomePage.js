import React from 'react';
import { history } from '../routers/AppRouter';

const onEnterClick = () => {
  history.push('/portfolio');
};

const HomePage = () => (
  <div className={'banner-layout'}>
    <div className={'banner-layout__banner'}>
      <h1 className={'banner-layout__subtitle'}>Firstname Lastname</h1>
      <p className={'banner-layout__title'}>Hong Kong</p>
    </div>
    <div>
      <button onClick={onEnterClick} className={'button button--larger'}>Enter</button>
    </div>
  </div>
);

export default HomePage;
