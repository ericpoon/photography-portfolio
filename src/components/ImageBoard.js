import React from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import SortableContainer from './SortableContainer';

const ROW_NUM = 3;
const COL_NUM = 3;

// todo: Item states are not persisted,
// use redux store to keep track of those images being edited (not all)

class ImageBoard extends React.Component {
  constructor(props) {
    super();
    this.state = { images: props.children };
  }

  componentWillReceiveProps(nextProps, nextContext) {
    this.setState({ images: nextProps.children });
  }

  moveCard = (dragIndex, hoverIndex) => {
    const { images } = this.state;
    const draggedCard = images[dragIndex];
    images.splice(dragIndex, 1, images[hoverIndex]);
    images.splice(hoverIndex, 1, draggedCard);
    this.setState({ images });
  };

  render() {
    const { images } = this.state;
    const { sortable } = this.props;
    const grid = [];

    images.forEach((current, index) => {
      const row = Math.floor(index / ROW_NUM);
      const col = index % COL_NUM;
      if (!grid[row]) grid[row] = [];
      grid[row][col] = sortable ? (
        <SortableContainer
          moveCard={this.moveCard}
          index={index}
          key={current.key}
        >
          {current}
        </SortableContainer>
      ) : current;
    });

    return (
      <div className={'image-card-collection'}>
        {grid.map(row => (
          <div className={'image-card-collection__row'} key={row.map(i => i.key).toString()}>
            {row.map(individual => individual)}
          </div>
        ))}
      </div>
    );
  }
}

export default DragDropContext(HTML5Backend)(ImageBoard);
