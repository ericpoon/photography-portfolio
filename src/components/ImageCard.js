import React from 'react';

class ImageCard extends React.Component {
  constructor(props) {
    super();
    const {
      title = '', subtitle = '', description = '', file = null, fileUrl = '',
    } = props;
    this.state = {
      title, subtitle, description, file, fileUrl,
    };
  }

  onTitleChange = (e) => {
    this.setState({ title: e.target.value.trim() });
  };
  onSubtitleChange = (e) => {
    this.setState({ subtitle: e.target.value.trim() });
  };
  onDescriptionChange = (e) => {
    this.setState({ description: e.target.value.trim() });
  };
  onFileChange = (file) => {
    console.log(file);
    this.setState({ file });
  };
  onSaveClick = () => {
    const {
      title, subtitle, description, file,
    } = this.state;
    const { onSaveClick } = this.props;
    const updates = {
      title, subtitle, description, file,
    };
    if (onSaveClick) onSaveClick(updates);
  };
  onDeleteClick = () => {
    this.setState({ title: '', subtitle: '', description: '' });
    const { onDeleteClick } = this.props;
    if (onDeleteClick) onDeleteClick();
  };
  onDiscardClick = () => {
    const { title = '', subtitle = '', description = '' } = this.props;
    this.setState({ title, subtitle, description });
  };

  isDirty = (field) => {
    const propVal = this.props[field] ? this.props[field] : '';
    const stateVal = this.state[field] ? this.state[field] : '';
    return propVal !== stateVal;
  };

  render() {
    const { onImageClick, editable } = this.props;
    const { title, subtitle, description } = this.state;
    if (editable) {
      return (
        <div className={'image-card'}>
          <div
            className={'image-card__image'}
            style={{ backgroundImage: `url(${this.state.fileUrl})` }}
            onClick={onImageClick}
          >
            <button
              className={'button button--floating-top-left button--danger'}
              onClick={this.onDeleteClick}
            />
          </div>
          <input type={'file'} onChange={e => this.onFileChange(e.target.files[0])} />
          <div>
            <input
              className={
                `image-card__title image-card__title--input${
                  this.isDirty('title') ? ' is-dirty' : ''}`
              }
              value={title}
              onChange={this.onTitleChange}
            />
            <input
              className={
                `image-card__subtitle image-card__subtitle--input${
                  this.isDirty('subtitle') ? ' is-dirty' : ''}`
              }
              value={subtitle}
              onChange={this.onSubtitleChange}
            />
            <textarea
              className={
                `image-card__description image-card__description--textarea${
                  this.isDirty('description') ? ' is-dirty' : ''}`
              }
              value={description}
              onChange={this.onDescriptionChange}
            />
            <div className={'image-card__actions'}>
              <button className={'button'} onClick={this.onSaveClick}>Save</button>
              <button className={'button'} onClick={this.onDiscardClick}>Discard</button>
            </div>
          </div>
        </div>
      );
    }

    return (
      <div className={'image-card'}>
        <div
          className={'image-card__image'}
          style={{ backgroundImage: `url(${this.state.fileUrl})` }}
          onClick={onImageClick}
        />
        <div>
          <h3 className={'image-card__title'}>{title}</h3>
          <h4 className={'image-card__subtitle'}>{subtitle}</h4>
          <p className={'image-card__description'}>{description}</p>
        </div>
      </div>
    );
  }
}

export default ImageCard;
