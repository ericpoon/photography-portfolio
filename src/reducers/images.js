const initialState = [];

export default (state = initialState, action) => {
  switch (action.type) {
    case 'SET_IMAGES':
      return action.images;
    case 'ADD_IMAGE':
      return [...state, action.image];
    case 'EDIT_IMAGE':
      return state.map((i) => {
        return i.id === action.id ? { ...i, ...action.updates } : i;
      });
    case 'DELETE_IMAGE':
      return state.filter(i => i.id !== action.id);
    default:
      return state;
  }
};
