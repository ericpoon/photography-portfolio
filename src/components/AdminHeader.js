import React from 'react';
import { connect } from 'react-redux';
import Header from './presentational/Header';
import { startLogout } from '../actions/auth';

export const AdminHeader = (props) => {
  return (
    <Header title={'Portfolio'}>
      <button
        onClick={() => null}
        className={'button button--no-background'}
      >
        {props.userName}
      </button>

      <button
        onClick={props.startLogout}
        className={'button button--no-background'}
      >
        Logout
      </button>
    </Header>
  );
};

const mapStateToProps = state => ({
  userName: state.auth.name,
});

export default connect(mapStateToProps, { startLogout })(AdminHeader);
