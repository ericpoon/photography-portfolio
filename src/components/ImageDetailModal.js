import React from 'react';
import { connect } from 'react-redux';
import Modal from 'react-modal';
import { hideDetail } from '../actions/gallery';

export const ImageDetailModal = (props) => {
  const { title, subtitle, description } = props;

  return (
    <Modal
      isOpen={props.isOpen}
      onRequestClose={() => props.hideDetail()}
      appElement={document.getElementById('app')}
    >
      <div>
        <img src={'/images/bg.jpg'} alt={''} />
        <h3>{title}</h3>
        <h4>{subtitle}</h4>
        <p>{description}</p>
      </div>
    </Modal>

  );
};

const mapStateToProps = (state) => {
  const image = state.images.filter(i => i.id === state.gallery.selectedId)[0];
  return {
    isOpen: state.gallery.showDetail,
    title: image ? image.title : null,
    subtitle: image ? image.subtitle : null,
    description: image ? image.description : null,
  };
};

export default connect(mapStateToProps, { hideDetail })(ImageDetailModal);
