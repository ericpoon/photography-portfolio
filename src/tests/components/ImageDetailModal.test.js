import React from 'react';
import { shallow } from 'enzyme';
import { ImageDetailModal } from '../../components/ImageDetailModal';
import images from '../fixtures/images';

let wrapper,
  hideDetail;

beforeEach(() => {
  const { title, subtitle, description } = images[0];
  hideDetail = jest.fn();
  wrapper = shallow(<ImageDetailModal
    isOpen
    title={title}
    subtitle={subtitle}
    description={description}
    hideDetail={hideDetail}
  />);
});

it('should render ImageDetailModal correctly when isOpen is true', () => {
  const modal = wrapper.find('Modal');
  expect(wrapper).toMatchSnapshot();
  expect(modal.prop('isOpen')).toBeTruthy();
});

it('should render ImageDetailModal correctly when isOpen is false', () => {
  wrapper.setProps({ isOpen: false });
  const modal = wrapper.find('Modal');
  expect(wrapper).toMatchSnapshot();
  expect(modal.prop('isOpen')).toBeFalsy();
});

it('should call hideDetail when closing', () => {
  const modal = wrapper.find('Modal');
  modal.prop('onRequestClose')();
  expect(hideDetail).toHaveBeenCalled();
});
