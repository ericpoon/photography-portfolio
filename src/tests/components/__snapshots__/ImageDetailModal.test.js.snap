// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ImageDetailModal correctly when isOpen is false 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={0}
  isOpen={false}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <img
      alt=""
    />
    <h3>
      Kowloon
    </h3>
    <h4>
      Dec 31, 2017
    </h4>
    <p>
      Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur
    </p>
  </div>
</Modal>
`;

exports[`should render ImageDetailModal correctly when isOpen is true 1`] = `
<Modal
  appElement={null}
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  closeTimeoutMS={0}
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnEsc={true}
  shouldCloseOnOverlayClick={true}
  shouldFocusAfterRender={true}
  shouldReturnFocusAfterClose={true}
>
  <div>
    <img
      alt=""
    />
    <h3>
      Kowloon
    </h3>
    <h4>
      Dec 31, 2017
    </h4>
    <p>
      Putarem quodque tamquam ii ob an deumque. Me assequi sapores ab verarum. Mei qui religionis nam fas devenietur ponderibus. Junctae invenit plausum differt de ipsamet externo is. Mutationum sui negationem exponuntur
    </p>
  </div>
</Modal>
`;
